crates/kitsune_p2p/kitsune_p2p/src/gossip/sharded_gossip/store.rs

```diff
@@ -0,0 +1,226 @@
+        .await?
+        .into_iter()
+        .filter(move |info| set.contains(info.agent.as_ref())))
+}
+
+/// Get agents and their intervals within a `DhtArcSet`.
+pub(super) async fn agents_within_arcset(
+    evt_sender: &EventSender,
+    space: &Arc<KitsuneSpace>,
+    arc_set: Arc<DhtArcSet>,
+) -> KitsuneResult<Vec<(Arc<KitsuneAgent>, ArcInterval)>> {
+    Ok(evt_sender
+        .query_gossip_agents(QueryGossipAgentsEvt {
+            space: space.clone(),
+            agents: None,
+            since_ms: 0,
+            until_ms: u64::MAX,
+            arc_set,
+        })
+        .await
+        .map_err(KitsuneError::other)?)
+}
+
+/// Get all ops for all agents that fall within the specified arcset.
+pub(super) async fn all_op_hashes_within_arcset(
+    evt_sender: &EventSender,
+    space: &Arc<KitsuneSpace>,
+    agents: &[(Arc<KitsuneAgent>, ArcInterval)],
+    common_arc_set: &DhtArcSet,
+    window_ms: TimeWindowMs,
+    max_ops: usize,
+    include_limbo: bool,
+) -> KitsuneResult<Option<(Vec<Arc<KitsuneOpHash>>, Range<u64>)>> {
+    let agents: Vec<_> = agents
+        .iter()
+        .map(|(a, i)| {
+            // Intersect this agent's arc with the arcset to find the minimal
+            // arcset relevant to this agent
+            let intersection = common_arc_set.intersection(&DhtArcSet::from_interval(i));
+            (a.clone(), intersection)
+        })
+        .collect();
+    Ok(evt_sender
+        .fetch_op_hashes_for_constraints(FetchOpHashesForConstraintsEvt {
+            space: space.clone(),
+            agents,
+            window_ms,
+            max_ops,
+            include_limbo,
+        })
+        .await
+        .map_err(KitsuneError::other)?)
+}
+
+/// Add new agent info to the p2p store.
+pub(super) async fn put_agent_info(
+    evt_sender: &EventSender,
+    space: &Arc<KitsuneSpace>,
+    agents_within_common_arc: HashSet<Arc<KitsuneAgent>>,
+    agents: &[Arc<AgentInfoSigned>],
+) -> KitsuneResult<()> {
+    for this_agent_info in all_agent_info(evt_sender, space)
+        .await?
+        .into_iter()
+        .filter(|a| agents_within_common_arc.contains(a.agent.as_ref()))
+    {
+        let peer_data = agents
+            .iter()
+            .filter(|new_info| {
+                this_agent_info
+                    .storage_arc
+                    .contains(new_info.agent.get_loc())
+            })
+            .map(|i| (**i).clone())
+            .collect();
+        evt_sender
+            .put_agent_info_signed(PutAgentInfoSignedEvt {
+                space: space.clone(),
+                peer_data,
+            })
+            .await
+            .map_err(KitsuneError::other)?;
+    }
+    Ok(())
+}
+
+pub(super) async fn fetch_ops(
+    evt_sender: &EventSender,
+    space: &Arc<KitsuneSpace>,
+    agents: impl Iterator<Item = &Arc<KitsuneAgent>>,
+    op_hashes: Vec<Arc<KitsuneOpHash>>,
+) -> KitsuneResult<Vec<(Arc<KitsuneOpHash>, Vec<u8>)>> {
```

Comment:
Seems silly to put the op hash in an Arc and not the op data itself, which is larger. Seems to speak to the question of if whether want to always put our kitsune hash types in Arcs, or if we want to do it case-by-case.

Reply:
We could put the data in an Arc too... but originally the hashes were in various HashSets so they actually got cloned, where the data was not. Once simple_bloom started keeping that sparse data map, perhaps it would have made sense to put an Arc around the data, because it was getting cloned, but the new gossip doesn't have a data map, right? so the data never gets cloned, and the Arc wouldn't help.

Reply:
Yeh I don't think we should put this in an arc